Update to gcc-8-20171210

belka-ew opened the pull request.

belka-ew committed:
Update to gcc-8-20171203

belka-ew commented:
The test that fails is 13089 from runnable/sdtor.d:
```d
extern (C) int printf(const(char*) fmt, ...);
/**********************************/
// 13089

struct S13089
{
    @disable this(this);    // non nothrow
}

void* p13089;

S13089[1000] foo13089() nothrow
{
    typeof(return) data;
    p13089 = &data;
    return data;
}

void test13089()
{
    immutable data = foo13089();
    assert(p13089 == &data);
}


int main()
{
    test13089();

    printf("Success\n");
    return 0;
}
```
Looking further...

ibuclaw commented:
This is the change that introduced the regression by the looks.

https://gcc.gnu.org/viewcvs/gcc?limit_changes=0&view=revision&revision=255066

ibuclaw commented:
The failure seems correct to me.  The struct is empty (`TYPE_EMPTY_P`), and so does not trigger NRVO code path.

ibuclaw commented:
There's an introduction of `DECL_PADDING_P` tree accessor in the current snapshot.  This could perhaps be set on fields generated by `build_alignment_field`.

belka-ew commented:
> This is the change that introduced the regression by the looks.
>
> https://gcc.gnu.org/viewcvs/gcc?limit_changes=0&view=revision&revision=255066

Yes, it is correct. The relevant change is in [gcc/function.c: aggregate_value_p](https://gcc.gnu.org/viewcvs/gcc/trunk/gcc/function.c?limit_changes=0&view=markup&pathrev=255066#l2087):
```cpp
  if (TYPE_EMPTY_P (type))
    return 0;
```
Removing the if-block fixes this test.

> The failure seems correct to me. The struct is empty (TYPE_EMPTY_P), and so does not trigger NRVO code path.

The code runs with dmd without an assertion failure.

> There's an introduction of DECL_PADDING_P tree accessor in the current snapshot. This could perhaps be set on fields generated by build_alignment_field.

Is it as simple as puting
```cpp
DECL_PADDING_P (field) = 1;
```
into `build_alignment_field`? It didn't work.

ibuclaw commented:
The macro accessor won't fix the issue. It's just something that we can begin to make use of.

belka-ew commented:
Btw. a lot of code replaced `size_in_bytes` with `arg_size_in_bytes`, probably something should be done in GDC as well.

ibuclaw commented:
In the past we've just fixed the D2 testsuite as needed. As there's normally a valid reason why dmd might be wrong. Such as returning data on the stack when gcc returns it in registers.

ibuclaw commented:
The NRV tests themselves are questionable because there is no guaranteed semantic on how it should work (it's an optimization, not a feature).

belka-ew commented:
I thought it is an optional feature only for C++, but something that is guaranteed for D.

ibuclaw commented:
> I thought it is an optional feature only for C++, but something that is guaranteed for D.

There's no sure fire way to determine whether NRVO happens or not in D.  It's not something you can just turn on for any function.

In the example given:
```
p13089 = &data;
```
What `p13089` points to after `data` has left scope is really undefined behaviour.  Doing pointer comparison as the testsuite is doing is making a very brave assumption that something specific will happen.

NRVO is not listed anywhere in the D language spec, only as part of the glossary:

https://dlang.org/glossary.html#nrvo

We don't lower code as described.  Of course if a precise semantic is required for NRVO that goes beyond what the GCC backend guarantees, then the NRVO code will need to be rewritten to be more in line with the above link.

belka-ew commented:
If NRVO isn't reliable should the test code compile at all? since the postblit of `S13089` is disabled, so a static array of `S13089` can't be copied.

There was a small discussion about RVO some years ago:
http://forum.dlang.org/post/gverkczeotvadwmdowdl@forum.dlang.org

belka-ew commented:
Maybe it doesn't matter for empty structs.

ibuclaw commented:
> so a static array of S13089 can't be copied.

AFAIU, `@disable this(this)` prevents copying, but doesn't prevent data being _moved_.

belka-ew commented:
just remove the test?

ibuclaw commented:
> just remove the test?

Or add a field to the struct so that its no longer empty.

ibuclaw commented:
Though we may have to consider what to do with all nrvo tests - or our implementation of nrvo - at some point.

Upstream also tests with `-fPIC`

https://github.com/dlang/dmd/blob/24de66934ff1412aec71ba5c5f6c2c3fa0f52664/test/runnable/sdtor.d#L1

If given the chance, the gcc optimizer will assume the condition `(p13089 == &data)` is always false, in this case turning on PIC prevents inlining.

belka-ew commented:
Should I open an issue, so it won't be forgotten?

ibuclaw commented:
Go for it.

belka-ew committed:
Fix runnable/sdtor for empty records

belka-ew commented:
I left `size_in_bytes` and `int_size_in_bytes` as is. The replacing causes:
```
FAIL: fail_compilation/ice13259.d   (internal compiler error)
FAIL: fail_compilation/ice13259.d -g   (internal compiler error)
FAIL: fail_compilation/ice13259.d -O2   (internal compiler error)
FAIL: fail_compilation/ice13259.d -g -O2   (internal compiler error)
FAIL: fail_compilation/ice13259.d -frelease   (internal compiler error)
FAIL: fail_compilation/ice13259.d -g -frelease   (internal compiler error)
FAIL: fail_compilation/ice13259.d -O2 -frelease   (internal compiler error)
FAIL: fail_compilation/ice13259.d -g -O2 -frelease   (internal compiler error)
```

ibuclaw commented:
Do you think you'll be able to diagnose why? Can always give tips.

belka-ew commented:
yes, I can look into it.

ibuclaw commented:
Having a look at the change history, there's a couple more things that may be of use to us.

- `SWITCH_ALL_CASES_P`: True if there are case labels for all possible values of SWITCH_COND, either because there is a default: case label or because the case label ranges cover all values.
- `SWITCH_BREAK_LABEL_P`: Set on the artificial label created for break; stmt from a switch.

The documentation of `arg_int_size_in_bytes` says: Like int_size_in_bytes, but handle empty records specially.

As the prefix suggests, it should only be used in conjunction with arguments (PARM_DECL), which may be nowhere in our codebase.  So concluding that it does not apply for us is still an acceptable decision.

belka-ew committed:
Use SWITCH_ALL_CASES_P and SWITCH_BREAK_LABEL_P

Set SWITCH_ALL_CASES_P on switch statements. Set SWITCH_BREAK_LABEL_P on the
artificial label created for break statements from a switch.

belka-ew commented:
As far as I understand `SWITCH_ALL_CASES_P` can be set for all D `switch` statements.
I'm not that sure about the position of `SWITCH_BREAK_LABEL_P`, I did my best by looking into C and C++ frontends.
If `arg_int_size_in_bytes` is not suitable for D, is it still interesting why and where it fails?

ibuclaw commented:
> If arg_int_size_in_bytes is not suitable for D, is it still interesting why and where it fails?

I imagine the why is because it returns `0`.  Where would be interesting only if we are able to reproduce the ICE by other means.

ibuclaw commented:
> As far as I understand SWITCH_ALL_CASES_P can be set for all D switch statements.

Right, as `default` must always be present, unless it's a `final switch` (in that case, the compiler inserts a default case for us).

ibuclaw commented:
> I'm not that sure about the position of SWITCH_BREAK_LABEL_P, I did my best by looking into C and C++ frontends.

Looks correct.  Though I had to turn the warning on explicitly for D in order to verify it. ;-)

https://github.com/gcc-mirror/gcc/blob/2422305dbe41bbc85956ff2208b4a820c3677d8f/gcc/gimplify.c#L2171-L2175

belka-ew commented:
> I imagine the why is because it returns 0.

It doesn't seem to be the case.

Internal error occurres in decl.cc, these lines:
```cpp
      dinteger_t tsize = arg_int_size_in_bytes (TREE_TYPE (decl));
      dinteger_t dtsize = arg_int_size_in_bytes (TREE_TYPE (DECL_INITIAL (decl)));
```
Only the second line fails. It has something to do with `DECL_INITIAL`. The error is:
```
output is fail_compilation/ice13259.d:8:11: error: non-constant nested delegate literal expression __dgliteral3
fail_compilation/ice13259.d:8:11: error: non-constant expression __dgliteral3
fail_compilation/ice13259.d:8:6: internal compiler error: tree check: expected class 'type', have 'exceptional' (error_mark) in arg_int_size_in_bytes, at tree.c:13903
0x696f48 ???
	../sysdeps/x86_64/start.S:118
```
The error is in `arg_int_size_in_bytes` itself, not in its return value (seems to me). `arg_int_size_in_bytes` is implemented as:
```cpp
tree
arg_size_in_bytes (const_tree type)
{
  return TYPE_EMPTY_P (type) ? size_zero_node : size_in_bytes (type);
}
```
Using `arg_size_in_bytes` in other files causes some unresolved testcases, but it could be my local problem, I have sometimes unresolved testcases, haven't looked further yet.

ibuclaw commented:
That is because `TYPE_EMPTY_P` expects to be given a type, not an `ERROR_MARK` node.

If you look at `size_in_bytes`, the first thing checked is:
```
if (type == error_mark_node)
  return integer_zero_node;
```

Perhaps these new functions should check for that as well, I can submit a patch to the gcc patches ml to see if it could be considered.

belka-ew commented:
Well, there are further errors coming from using `arg_*` functions.

```
libtool: compile:  /home/belka/GDC/build/./gcc/gdc -B/home/belka/GDC/build/./gcc/ -B/usr/local/x86_64-pc-linux-gnu/bin/ -B/usr/local/x86_64-pc-linux-gnu/lib/ -isystem /usr/local/x86_64-pc-linux-gnu/include -isystem /usr/local/x86_64-pc-linux-gnu/sys-include -Wall -Werror -g -fno-release -funittest -nostdinc -pipe -Wno-deprecated -I /home/belka/GDC/libphobos/src -I /home/belka/GDC/libphobos/libdruntime -I ../libdruntime -I . -c /home/belka/GDC/libphobos/src/std/conv.d  -fPIC -fversion=Shared -o std/.libs/conv.t.o
/home/belka/GDC/libphobos/src/std/conv.d: In function ‘emplaceInitializer’:
/home/belka/GDC/libphobos/src/std/conv.d:4378:28: internal compiler error: Mismatch between declaration '‘std.conv.emplaceInitializer!(std.conv.__unittestL4703_108.S1[2]).emplaceInitializer.init’' size (0) and it's initializer size (2).
         static immutable T init = T.init;
                            ^
0x79c386 d_finish_decl(tree_node*)
	/home/belka/GDC/gcc/d/decl.cc:1445
0x7a2253 DeclVisitor::visit(VarDeclaration*)
	/home/belka/GDC/gcc/d/decl.cc:611
0x79c151 build_decl_tree(Dsymbol*)
	/home/belka/GDC/gcc/d/decl.cc:898
0x7a37a1 ExprVisitor::visit(DeclarationExp*)
	/home/belka/GDC/gcc/d/expr.cc:2007
0x7a27ba build_expr(Expression*, bool)
	/home/belka/GDC/gcc/d/expr.cc:3026
0x7a2943 build_expr_dtor(Expression*)
	/home/belka/GDC/gcc/d/expr.cc:3071
0x7b2531 IRVisitor::visit(ExpStatement*)
	/home/belka/GDC/gcc/d/toir.cc:1008
0x7b2272 IRVisitor::build_stmt(Statement*)
	/home/belka/GDC/gcc/d/toir.cc:249
0x7b2272 IRVisitor::visit(CompoundStatement*)
	/home/belka/GDC/gcc/d/toir.cc:1025
0x7b2272 IRVisitor::build_stmt(Statement*)
	/home/belka/GDC/gcc/d/toir.cc:249
0x7b2272 IRVisitor::visit(CompoundStatement*)
	/home/belka/GDC/gcc/d/toir.cc:1025
0x7b1c1f IRVisitor::build_stmt(Statement*)
	/home/belka/GDC/gcc/d/toir.cc:249
0x7b1c1f build_function_body(FuncDeclaration*)
	/home/belka/GDC/gcc/d/toir.cc:1428
0x7a1515 DeclVisitor::visit(FuncDeclaration*)
	/home/belka/GDC/gcc/d/decl.cc:858
0x7a0c5f DeclVisitor::visit(TemplateInstance*)
	/home/belka/GDC/gcc/d/decl.cc:266
0x79c151 build_decl_tree(Dsymbol*)
	/home/belka/GDC/gcc/d/decl.cc:898
0x7ae5a0 build_module_tree(Module*)
	/home/belka/GDC/gcc/d/modules.cc:716
0x7a0b3b DeclVisitor::visit(Module*)
	/home/belka/GDC/gcc/d/decl.cc:137
0x79c151 build_decl_tree(Dsymbol*)
	/home/belka/GDC/gcc/d/decl.cc:898
0x7953a6 d_parse_file()
	/home/belka/GDC/gcc/d/d-lang.cc:1224
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <http://bugzilla.gdcproject.org> for instructions.
Makefile:1796: recipe for target 'std/conv.t.lo' failed
make[3]: *** [std/conv.t.lo] Error 1
```

ibuclaw commented:
And that assert is very deliberate. What it's testing is that for the given static variable we are sending to backend, it's initializer is the same size as the declaration type.

This catches early a bug/ice in the arm backend when under optimizations, where section anchors broke if the initializer was larger than the declaration (this used to happen for Typeinfo_Class data, where overridden interfaces are written immediately after the symbol).

Maybe we could skip generating an initializer for empty variables, as it would only be zero padding anyway.

As for using the `arg_*` functions, I still think that they only make sense when getting the ABI size of parameters.

belka-ew commented:
I rebased on master. If you want you can leave this unmerged. I try to build it with the new snapshot in the next days. Otherwise I'll create another PR.

belka-ew committed:
Update to gcc-8-20171210
